### PR TITLE
[ONNX] Fix arg type in `_set_training_mode`

### PR DESCRIPTION
**Pull request by justinchuby**
When `TrainingMode.PRESERVE` is set for export, the exporter used to change the model's training mode based on some logic. Now we respect the option and not touch the model's training state.

- Previously `_set_training_mode`'s behavior doesn't match what the global variable expects. This PR removes the deprecated `_set_training_mode` and makes the type correct.

### CODE REVIEW

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -46,48 +46,58 @@ def is_in_onnx_export():
...
         yield
     finally:
-        if not isinstance(model, torch.jit.ScriptFunction):
-            # FIXME(justinchuby): is_originally_training is possibly unbound
-            model.train(is_originally_training)
+        if not (
+            isinstance(model, torch.jit.ScriptFunction)
+            or export_mode == _C_onnx.TrainingMode.PRESERVE
+        ):
+            # Alter the model training state only if export_mode is not PRESERVE
```
Comment (BowenBao) [inline torch/onnx/utils.py:99]: nit: comment seems obvious and repeatetive to above code

File: torch/onnx/symbolic_opset10.py [hunk 2]
```diff
@@ -324,7 +324,7 @@ def embedding_bag(
     include_last_offset,
     padding_idx,
 ):
-    if scale_grad_by_freq and GLOBALS.training_mode:
+    if scale_grad_by_freq and GLOBALS.training_mode == _C_onnx.TrainingMode.TRAINING:
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset10.py:327]: will you revert `GLOBALS.training_mode` to boolean or keep it as enum? Current enum setting seems to have an issue with the case when `mode == PRESERVE and original_export is True`.

File: torch/onnx/_globals.py [hunk 3]
```diff
@@ -24,10 +24,24 @@ class _InternalGlobals:
 
     def __init__(self):
         self._export_onnx_opset_version = _constants.onnx_default_opset
+        self._training_mode: Optional[_C_onnx.TrainingMode] = None
```
Comment (BowenBao) [inline torch/onnx/_globals.py:27]: How about setting default to EVAL and drop the `Optional` (here and at the context manager)?

File: torch/onnx/symbolic_opset10.py (hunk 2 above)
Comment (justinchuby) [reply torch/onnx/symbolic_opset10.py:327]: Currently I am keeping it as enum. `mode == PRESERVE and originally_training is True` is the case I was concerned about. WIth this we treat them as EVAL mode even when originally_training is True. What do you think?

File: torch/onnx/_globals.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/_globals.py]: Done

File: torch/onnx/utils.py (hunk 1 above)
Comment (justinchuby) [reply torch/onnx/utils.py]: Done

File: torch/onnx/symbolic_opset10.py (hunk 2 above)
Comment (justinchuby) [reply torch/onnx/symbolic_opset10.py]: Updated

File: torch/onnx/verification.py [hunk 4]
```diff
@@ -506,7 +508,8 @@ def verify(
     ] = None,
     input_names: Optional[Sequence[str]] = None,
     output_names: Optional[Sequence[str]] = None,
-    training: Optional[bool] = None,
+    # TODO(justinchuby): Default training = torch.onnx.TrainingMode.EVAL after #78446 is checked in
```
Comment (BowenBao) [inline torch/onnx/verification.py:511]: nit: this is done

**Review by BowenBao — APPROVED** [review]
LGTM!

File: torch/onnx/verification.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/verification.py:511]: Good catch, done!
